### PR TITLE
Delay node: improvements

### PR DESCRIPTION
**Pull request by irynamatveieva**
## Pull Request description

- added tests for the delay node
- changed configuration to allow set time unit using patterns
- provided upgrade for the existing nodes
- cover new functionality with tests

PE PR: https://github.com/thingsboard/thingsboard-pe/pull/2834

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [x] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

**Comment by dskarzh** [comment]
- Use concurrent hash map instead of normal one, because there is not guarantee that map will always be accessed by the same thread
- Make time unit an enum
- Add correct validation for all configuration fields (period not null, time unit not null and after parsing it is one of the supported values meaning it should not be possible to use microseconds for example, max pending messages is positive and not greater than 100, etc, please consult UI to check how it validates)

**Review by dskarzh — APPROVED** [review]

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/delay/TbMsgDelayNodeTest.java [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+                                    }
+                                """,
+                        true,
+                        """
+                                    {
+                                        "period": "60",
+                                        "timeUnit": "SECONDS",
+                                        "maxPendingMsgs": 1000
+                                    }
+                                """
```
Comment (ShvaykaD) [inline rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/delay/TbMsgDelayNodeTest.java:254]: why period was changed from 13 seconds to 60 seconds?

**Review by ShvaykaD — CHANGES_REQUESTED** [review]

File: rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/delay/TbMsgDelayNodeTest.java (hunk 1 above)
Comment (ShvaykaD) [reply rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/delay/TbMsgDelayNodeTest.java:254]: > **If there is no useMetadataPeriodInSecondsPatterns then set period to "60"**

This feature didn't exist at the time the rule node was created. So potentially there might be cases where this parameter doesn't exist in the rule node configuration. Non-existing boolean parameters are treated as false during the parse of json config to the configuration class..

**Review by ShvaykaD — APPROVED** [review]

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/delay/TbMsgDelayNode.java [hunk 2]
```diff
@@ -50,13 +56,17 @@
 )
 public class TbMsgDelayNode implements TbNode {
 
+    private static final Set<TimeUnit> supportedTimeUnits = EnumSet.of(TimeUnit.SECONDS, TimeUnit.MINUTES, TimeUnit.HOURS);
+    private static final String supportedTimeUnitsStr = String.join(",", TimeUnit.SECONDS.name(), TimeUnit.MINUTES.name(), TimeUnit.HOURS.name());
```
Comment (ViacheslavKlimov) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/delay/TbMsgDelayNode.java:60]: Join supportedTimeUnits.
Also, add a space to "," separator

File: rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/delay/TbMsgDelayNode.java [hunk 3]
```diff
@@ -89,25 +99,81 @@ public void onMsg(TbContext ctx, TbMsg msg) {
...
+    private void validateConfig() throws TbNodeException {
+        if (config.getMaxPendingMsgs() < 1 || config.getMaxPendingMsgs() > 100000) {
+            throw new TbNodeException("Maximum pending messages should be in a range from 1 to 100000.", true);
+        }
+        if (config.getPeriod() == null) {
+            throw new TbNodeException("Period should be specified.", true);
+        }
+        if (config.getTimeUnit() == null) {
+            throw new TbNodeException("Time unit should be specified.", true);
+        }
```
Comment (ViacheslavKlimov) [inline rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/delay/TbMsgDelayNode.java:128]: Please use javax constraints.

**Review by ViacheslavKlimov — CHANGES_REQUESTED** [review]

**Review by ViacheslavKlimov — APPROVED** [review]

**Pull request merged**